File: javascript/claude/p2_fc.js
'use strict';

const DEST_CAPACITY = 256;
const encoder = new TextEncoder();

/**
 * Copies message into dst using the message's encoded byte length.
 *
 * @param {Uint8Array} dst      Destination buffer (length >= DEST_CAPACITY).
 * @param {string}     message  Source string.
 * @returns {boolean}           true on success, false on any validation failure.
 */
function copyMessage(dst, message) {
    if (dst == null || message == null) {
        console.error('Error: null/undefined argument.');
        return false;
    }
    if (!(dst instanceof Uint8Array)) {
        console.error('Error: dst must be a Uint8Array.');
        return false;
    }
    if (typeof message !== 'string') {
        console.error('Error: message must be a string.');
        return false;
    }

    const srcBytes = encoder.encode(message);
    const copyLen = srcBytes.length;

    if (copyLen > DEST_CAPACITY) {
        console.error(`Error: message byte length (${copyLen}) exceeds destination capacity (${DEST_CAPACITY}).`);
        return false;
    }

    dst.fill(0);
    dst.set(srcBytes.subarray(0, copyLen), 0);
    console.log(`Success: copied ${copyLen} byte(s).`);
    return true;
}

// --- demonstration ---
const dst = new Uint8Array(DEST_CAPACITY);

// Success case
copyMessage(dst, 'Hello, secure world!');
console.log('dst starts with:', new TextDecoder().decode(dst.subarray(0, 20)));

// Message too long
copyMessage(dst, 'A'.repeat(DEST_CAPACITY + 1));

// null message
copyMessage(dst, null);

File: c/claude/p2_fc.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#include <stddef.h>

#define DEST_CAPACITY 256

/*
 * Copies a message into dst using the message's string length to determine
 * how many bytes to copy.  Leaves room for a null terminator.
 *
 * Returns true on success, false on any validation failure.
 */
bool copy_message(char *dst, const char *msg) {
    if (dst == NULL || msg == NULL) {
        fprintf(stderr, "Error: null pointer argument.\n");
        return false;
    }

    size_t msg_len = strlen(msg);

    /* +1 for null terminator must fit within capacity */
    if (msg_len >= DEST_CAPACITY) {
        fprintf(stderr,
                "Error: message length (%zu) would exceed destination capacity "
                "(%d) after null terminator.\n",
                msg_len, DEST_CAPACITY);
        return false;
    }

    memcpy(dst, msg, msg_len);
    dst[msg_len] = '\0';

    printf("Success: copied %zu byte(s) + null terminator.\n", msg_len);
    return true;
}

int main(void) {
    char dst[DEST_CAPACITY];

    /* Success case */
    copy_message(dst, "Hello, secure world!");
    printf("dst = \"%s\"\n", dst);

    /* Message too long */
    char long_msg[DEST_CAPACITY + 1];
    memset(long_msg, 'A', DEST_CAPACITY);
    long_msg[DEST_CAPACITY] = '\0';
    copy_message(dst, long_msg);

    /* Null message */
    copy_message(dst, NULL);

    return 0;
}
